src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +64,274 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
+
+	StringInfo selectQuery = makeStringInfo();
+	ListCell *shardCell = NULL;
+
+	appendStringInfo(selectQuery, "SELECT ");
+
+
+	foreach(shardCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardCell);
+		uint64 shardId = shardInterval->shardId;
+		char *shardName = get_rel_name(distributedRelationId);
+		char *shardQualifiedName = NULL;
+		char *quotedShardName = NULL;
+		AppendShardIdToName(&shardName, shardId);
+
+		shardQualifiedName = quote_qualified_identifier(schemaName, shardName);
+		quotedShardName = quote_literal_cstr(shardQualifiedName);
+
+		appendStringInfo(selectQuery, sizeQuery, quotedShardName);
+		appendStringInfo(selectQuery, " + ");
+	}
+
+	/* there is no more size function so just add 0 instead of it */
+	appendStringInfo(selectQuery, "0;");
+	return selectQuery;
+}
+
+
+/*
+ * SupportFunctionForCitusSize is a helper function for all citus size functions. It
+ * first checks whether the table is distributed and size query can be run on it.
+ * Connection to each node has to be established to get the size of the table.
+ */
+uint64
+SupportFunctionForCitusSize(Oid relationId, char *sizeQuery)
+{
+	MultiConnection *connection = NULL;
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+	uint64 totalRelationSize = 0;
+
+	/* check whether the size of table can be obtained with citus size functions */
+	ErrorIfNotSuitableToGetSize(relationId);
+
+	/* accessing single placement is enough for reference tables */
+	if (PartitionMethod(relationId) == DISTRIBUTE_BY_NONE)
+	{
+		PGresult *queryResult = NULL;
+		StringInfo tableSizeQuery = NULL;
+		ListCell *referenceWorkerNodeCell = list_head(workerNodeList);
+		WorkerNode *referenceWorkerNode = (WorkerNode *) lfirst(referenceWorkerNodeCell);
+
+		char *referenceWorkerNodeName = referenceWorkerNode->workerName;
+		uint32 referenceWorkerNodePort = referenceWorkerNode->workerPort;
+		char *tableSizeString;
+		int executeCommand = 0;
+		uint32 connectionFlag = 0;
+		uint64 currentRelationSize = 0;
+
+		List *intervalsOnNode = GetShardIntervalsOnNode(referenceWorkerNode, relationId);
+
+		tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(relationId,
+															   intervalsOnNode,
+															   sizeQuery);
+
+		connection = GetNodeConnection(connectionFlag,
+									   referenceWorkerNodeName,
+									   referenceWorkerNodePort);
+
+		executeCommand = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+													  &queryResult);
+
+		/* If there is a connection error close the connection and quit */
+		if (executeCommand != 0)
+		{
+			ReportConnectionError(connection, ERROR);
+		}
+
+		tableSizeString = PQgetvalue(queryResult, 0, 0);
+		currentRelationSize = atol(tableSizeString);
+
+		PQclear(queryResult);
+
+		ForgetResults(connection);
+
+		totalRelationSize = list_length(workerNodeList) * currentRelationSize;
+		return totalRelationSize;
+	}
+
+	foreach(workerNodeCell, workerNodeList)
```

Comment:
Yes, the loop works for each kind of distributed table independent of replication factor. 

Since we know that same values are held on each worker node for the reference tables, we just sum them up to get final table size. If we include the logic of reference table in that loop, we will unnecessarily loop over each nodes although getting the size from single node is enough. Yet, we can change the implementation as you mentioned since our loss will be O(nodeCount), which can be compensated by code quality.